### PR TITLE
arm/riscv/xtensa: enhance the task dump 

### PR DESCRIPTION
**Pull request by anchao**
## Summary

arm/riscv/xtensa: enhance the task dump

add irq stack information
add cpu loading

## Impact

N/A, assert dump

## Testing

before:

```
Idle Task: PID=0 PRI=0 Stack Used=512 of 3048
hpwork: PID=1 PRI=224 Stack Used=304 of 2016
lpwork: PID=2 PRI=100 Stack Used=304 of 2016
rptun: PID=4 PRI=224 Stack Used=856 of 2008
```

after:

```
[ EMERG] [ap] arm_showtasks:    PID    PRI      USED     STACK   FILLED       CPU   COMMAND
[ EMERG] [ap] arm_showtasks:   ----   ----       928      2048    45.3%      ----   irq
[ EMERG] [ap] arm_dump_task:      0      0       512      3048    16.7%     99.4%   Idle Task
[ EMERG] [ap] arm_dump_task:      1    224       304      2016    15.0%      0.0%   hpwork
[ EMERG] [ap] arm_dump_task:      2    100       304      2016    15.0%      0.0%   lpwork
[ EMERG] [ap] arm_dump_task:      4    224       856      2008    42.6%      0.0%   rptun

```

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/risc-v/src/rv64gc/riscv_assert.c [hunk 1]
```diff
@@ -132,63 +132,171 @@ static inline void riscv_registerdump(volatile uintptr_t *regs)
 }
 
 /****************************************************************************
- * Name: up_taskdump
+ * Name: riscv_dump_stak
```
Comment (gustavonihei) [inline arch/risc-v/src/rv64gc/riscv_assert.c:135]:
```suggestion
 * Name: riscv_dump_task
```
Typo in the function name.

File: arch/risc-v/src/rv64gc/riscv_assert.c [hunk 2]
```diff
@@ -132,63 +132,171 @@ static inline void riscv_registerdump(volatile uintptr_t *regs)
...
  ****************************************************************************/
 
-#if defined(CONFIG_STACK_COLORATION) || defined(CONFIG_SCHED_BACKTRACE)
-static void up_taskdump(struct tcb_s *tcb, void *arg)
+static void riscv_dump_stak(FAR struct tcb_s *tcb, FAR void *arg)
```
Comment (gustavonihei) [inline arch/risc-v/src/rv64gc/riscv_assert.c:138]:
```suggestion
static void riscv_dump_task(FAR struct tcb_s *tcb, FAR void *arg)
```

File: arch/xtensa/src/common/xtensa_dumpstate.c [hunk 3]
```diff
@@ -53,47 +53,165 @@ static uint32_t s_last_regs[XCPTCONTEXT_REGS];
  ****************************************************************************/
 
 /****************************************************************************
- * Name: up_taskdump
+ * Name: xtensa_dump_stak
```
Comment (gustavonihei) [inline arch/xtensa/src/common/xtensa_dumpstate.c:56]:
```suggestion
 * Name: xtensa_dump_task
```
Typo in the function name.

File: arch/xtensa/src/common/xtensa_dumpstate.c [hunk 4]
```diff
@@ -53,47 +53,165 @@ static uint32_t s_last_regs[XCPTCONTEXT_REGS];
...
  ****************************************************************************/
 
-#ifdef CONFIG_STACK_COLORATION
-static void up_taskdump(struct tcb_s *tcb, void *arg)
+static void xtensa_dump_stak(FAR struct tcb_s *tcb, FAR void *arg)
```
Comment (gustavonihei) [inline arch/xtensa/src/common/xtensa_dumpstate.c:59]:
```suggestion
static void xtensa_dump_task(FAR struct tcb_s *tcb, FAR void *arg)
```

File: arch/risc-v/src/rv64gc/riscv_assert.c [hunk 5]
```diff
@@ -132,63 +132,171 @@ static inline void riscv_registerdump(volatile uintptr_t *regs)
...
 
 /****************************************************************************
- * Name: up_showtasks
+ * Name: riscv_showtasks
  ****************************************************************************/
 
-static inline void up_showtasks(void)
+static inline void riscv_showtasks(void)
 {
+#if CONFIG_ARCH_INTERRUPTSTACK > 7
```
Comment (gustavonihei) [inline arch/risc-v/src/rv64gc/riscv_assert.c:229]:
```suggestion
#if CONFIG_ARCH_INTERRUPTSTACK > 15
```
RISC-V requires a 16-byte stack alignment.

File: arch/risc-v/src/rv64gc/riscv_assert.c [hunk 6]
```diff
@@ -132,63 +132,171 @@ static inline void riscv_registerdump(volatile uintptr_t *regs)
...
+#  ifdef CONFIG_STACK_COLORATION
+  uint32_t stack_used = up_check_intstack();
+  uint32_t stack_filled = 0;
+
+  if ((CONFIG_ARCH_INTERRUPTSTACK & ~7) > 0 && stack_used > 0)
```
Comment (gustavonihei) [inline arch/risc-v/src/rv64gc/riscv_assert.c:234]:
```suggestion
  if ((CONFIG_ARCH_INTERRUPTSTACK & ~15) > 0 && stack_used > 0)
```

File: arch/risc-v/src/rv64gc/riscv_assert.c [hunk 7]
```diff
@@ -132,63 +132,171 @@ static inline void riscv_registerdump(volatile uintptr_t *regs)
...
+#endif
+#ifdef CONFIG_SCHED_CPULOAD
+         "      CPU"
+#endif
+#if CONFIG_TASK_NAME_SIZE > 0
+         "   COMMAND"
+#endif
+         "\n");
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 7
```
Comment (gustavonihei) [inline arch/risc-v/src/rv64gc/riscv_assert.c:262]:
```suggestion
#if CONFIG_ARCH_INTERRUPTSTACK > 15
```

File: arch/risc-v/src/rv64gc/riscv_assert.c [hunk 8]
```diff
@@ -132,63 +132,171 @@ static inline void riscv_registerdump(volatile uintptr_t *regs)
...
+         "     ----"
+#  endif
+#  if CONFIG_TASK_NAME_SIZE > 0
+         "   irq"
+#  endif
+         "\n"
+#  ifdef CONFIG_STACK_COLORATION
+         , (unsigned long)stack_used
+#  endif
+         , (unsigned long)(CONFIG_ARCH_INTERRUPTSTACK & ~7)
```
Comment (gustavonihei) [inline arch/risc-v/src/rv64gc/riscv_assert.c:281]:
```suggestion
         , (unsigned long)(CONFIG_ARCH_INTERRUPTSTACK & ~15)
```

File: arch/xtensa/src/common/xtensa_dumpstate.c [hunk 9]
```diff
@@ -53,47 +53,165 @@ static uint32_t s_last_regs[XCPTCONTEXT_REGS];
...
 /****************************************************************************
- * Name: up_showtasks
+ * Name: xtensa_showtasks
  ****************************************************************************/
 
-#ifdef CONFIG_STACK_COLORATION
-static inline void up_showtasks(void)
+static inline void xtensa_showtasks(void)
 {
+#if CONFIG_ARCH_INTERRUPTSTACK > 7
```
Comment (gustavonihei) [inline arch/xtensa/src/common/xtensa_dumpstate.c:150]:
```suggestion
#if CONFIG_ARCH_INTERRUPTSTACK > 15
```
Xtensa also requires a 16-byte stack alignment.

File: arch/xtensa/src/common/xtensa_dumpstate.c [hunk 10]
```diff
@@ -53,47 +53,165 @@ static uint32_t s_last_regs[XCPTCONTEXT_REGS];
...
+#  ifdef CONFIG_STACK_COLORATION
+  uint32_t stack_used = up_check_intstack();
+  uint32_t stack_filled = 0;
+
+  if ((CONFIG_ARCH_INTERRUPTSTACK & ~7) > 0 && stack_used > 0)
```
Comment (gustavonihei) [inline arch/xtensa/src/common/xtensa_dumpstate.c:155]:
```suggestion
  if ((CONFIG_ARCH_INTERRUPTSTACK & ~15) > 0 && stack_used > 0)
```

File: arch/xtensa/src/common/xtensa_dumpstate.c [hunk 11]
```diff
@@ -53,47 +53,165 @@ static uint32_t s_last_regs[XCPTCONTEXT_REGS];
...
+    {
+      /* Use fixed-point math with one decimal place */
+
+      stack_filled = 10 * 100 *
+                     stack_used / (CONFIG_ARCH_INTERRUPTSTACK & ~7);
```
Comment (gustavonihei) [inline arch/xtensa/src/common/xtensa_dumpstate.c:160]:
```suggestion
                     stack_used / (CONFIG_ARCH_INTERRUPTSTACK & ~15);
```

File: arch/xtensa/src/common/xtensa_dumpstate.c [hunk 12]
```diff
@@ -53,47 +53,165 @@ static uint32_t s_last_regs[XCPTCONTEXT_REGS];
...
 #endif
+#ifdef CONFIG_SCHED_CPULOAD
+         "      CPU"
+#endif
+#if CONFIG_TASK_NAME_SIZE > 0
+         "   COMMAND"
+#endif
+         "\n");
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 7
```
Comment (gustavonihei) [inline arch/xtensa/src/common/xtensa_dumpstate.c:183]:
```suggestion
#if CONFIG_ARCH_INTERRUPTSTACK > 15
```

File: arch/risc-v/src/rv32im/riscv_assert.c [hunk 13]
```diff
@@ -123,57 +123,165 @@ static inline void riscv_registerdump(volatile uint32_t *regs)
 #endif
 
 /****************************************************************************
- * Name: riscv_taskdump
+ * Name: riscv_dump_stak
```
Comment (gustavonihei) [inline arch/risc-v/src/rv32im/riscv_assert.c:126]:
```suggestion
 * Name: riscv_dump_task
```

File: arch/risc-v/src/rv32im/riscv_assert.c [hunk 14]
```diff
@@ -123,57 +123,165 @@ static inline void riscv_registerdump(volatile uint32_t *regs)
...
  ****************************************************************************/
 
-#if defined(CONFIG_STACK_COLORATION) || defined(CONFIG_SCHED_BACKTRACE)
-static void riscv_taskdump(struct tcb_s *tcb, void *arg)
+static void riscv_dump_stak(FAR struct tcb_s *tcb, FAR void *arg)
```
Comment (gustavonihei) [inline arch/risc-v/src/rv32im/riscv_assert.c:129]:
```suggestion
static void riscv_dump_task(struct tcb_s *tcb, void *arg)
```

File: arch/risc-v/src/rv32im/riscv_assert.c [hunk 15]
```diff
@@ -123,57 +123,165 @@ static inline void riscv_registerdump(volatile uint32_t *regs)
...
 }
+#endif
 
 /****************************************************************************
  * Name: riscv_showtasks
  ****************************************************************************/
 
 static inline void riscv_showtasks(void)
 {
+#if CONFIG_ARCH_INTERRUPTSTACK > 7
```
Comment (gustavonihei) [inline arch/risc-v/src/rv32im/riscv_assert.c:220]:
```suggestion
#if CONFIG_ARCH_INTERRUPTSTACK > 15
```
RISC-V requires a 16-byte stack alignment.

File: arch/risc-v/src/rv32im/riscv_assert.c [hunk 16]
```diff
@@ -123,57 +123,165 @@ static inline void riscv_registerdump(volatile uint32_t *regs)
...
+#  ifdef CONFIG_STACK_COLORATION
+  uint32_t stack_used = up_check_intstack();
+  uint32_t stack_filled = 0;
+
+  if ((CONFIG_ARCH_INTERRUPTSTACK & ~7) > 0 && stack_used > 0)
```
Comment (gustavonihei) [inline arch/risc-v/src/rv32im/riscv_assert.c:225]:
```suggestion
  if ((CONFIG_ARCH_INTERRUPTSTACK & ~15) > 0 && stack_used > 0)
```

File: arch/risc-v/src/rv32im/riscv_assert.c [hunk 17]
```diff
@@ -123,57 +123,165 @@ static inline void riscv_registerdump(volatile uint32_t *regs)
...
+    {
+      /* Use fixed-point math with one decimal place */
+
+      stack_filled = 10 * 100 *
+                     stack_used / (CONFIG_ARCH_INTERRUPTSTACK & ~7);
```
Comment (gustavonihei) [inline arch/risc-v/src/rv32im/riscv_assert.c:230]:
```suggestion
                     stack_used / (CONFIG_ARCH_INTERRUPTSTACK & ~15);
```

File: arch/risc-v/src/rv32im/riscv_assert.c [hunk 18]
```diff
@@ -123,57 +123,165 @@ static inline void riscv_registerdump(volatile uint32_t *regs)
...
+#endif
+#ifdef CONFIG_SCHED_CPULOAD
+         "      CPU"
 #endif
+#if CONFIG_TASK_NAME_SIZE > 0
+         "   COMMAND"
+#endif
+         "\n");
+
+#if CONFIG_ARCH_INTERRUPTSTACK > 7
```
Comment (gustavonihei) [inline arch/risc-v/src/rv32im/riscv_assert.c:253]:
```suggestion
#if CONFIG_ARCH_INTERRUPTSTACK > 15
```

File: arch/risc-v/src/rv32im/riscv_assert.c [hunk 19]
```diff
@@ -123,57 +123,165 @@ static inline void riscv_registerdump(volatile uint32_t *regs)
...
+         "     ----"
+#  endif
+#  if CONFIG_TASK_NAME_SIZE > 0
+         "   irq"
+#  endif
+         "\n"
+#  ifdef CONFIG_STACK_COLORATION
+         , (unsigned long)stack_used
+#  endif
+         , (unsigned long)(CONFIG_ARCH_INTERRUPTSTACK & ~7)
```
Comment (gustavonihei) [inline arch/risc-v/src/rv32im/riscv_assert.c:272]:
```suggestion
         , (unsigned long)(CONFIG_ARCH_INTERRUPTSTACK & ~15)
```

File: arch/xtensa/src/common/xtensa_dumpstate.c [hunk 20]
```diff
@@ -53,47 +53,165 @@ static uint32_t s_last_regs[XCPTCONTEXT_REGS];
...
+         "     ----"
+#  endif
+#  if CONFIG_TASK_NAME_SIZE > 0
+         "   irq"
+#  endif
+         "\n"
+#  ifdef CONFIG_STACK_COLORATION
+         , (unsigned long)stack_used
+#  endif
+         , (unsigned long)(CONFIG_ARCH_INTERRUPTSTACK & ~7)
```
Comment (gustavonihei) [inline arch/xtensa/src/common/xtensa_dumpstate.c:202]:
```suggestion
         , (unsigned long)(CONFIG_ARCH_INTERRUPTSTACK & ~15)
```

File: arch/xtensa/src/common/xtensa_dumpstate.c [hunk 21]
```diff
@@ -53,47 +53,165 @@ static uint32_t s_last_regs[XCPTCONTEXT_REGS];
...
+#  ifdef CONFIG_STACK_COLORATION
+         , stack_filled / 10, stack_filled % 10,
+         (stack_filled >= 10 * 80 ? '!' : ' ')
+#  endif
+        );
+#endif
+
+  nxsched_foreach(xtensa_dump_stak, NULL);
```
Comment (gustavonihei) [inline arch/xtensa/src/common/xtensa_dumpstate.c:210]:
```suggestion
  nxsched_foreach(xtensa_dump_task, NULL);
```

File: arch/risc-v/src/rv32im/riscv_assert.c [hunk 22]
```diff
@@ -123,57 +123,165 @@ static inline void riscv_registerdump(volatile uint32_t *regs)
...
+#  ifdef CONFIG_STACK_COLORATION
+         , stack_filled / 10, stack_filled % 10,
+         (stack_filled >= 10 * 80 ? '!' : ' ')
+#  endif
+        );
+#endif
+
+  nxsched_foreach(riscv_dump_stak, NULL);
```
Comment (gustavonihei) [inline arch/risc-v/src/rv32im/riscv_assert.c:280]:
```suggestion
  nxsched_foreach(riscv_dump_task, NULL);
```

File: arch/risc-v/src/rv64gc/riscv_assert.c [hunk 23]
```diff
@@ -132,63 +132,171 @@ static inline void riscv_registerdump(volatile uintptr_t *regs)
...
+#  ifdef CONFIG_STACK_COLORATION
+         , stack_filled / 10, stack_filled % 10,
+         (stack_filled >= 10 * 80 ? '!' : ' ')
+#  endif
+        );
 #endif
 
+  nxsched_foreach(riscv_dump_stak, NULL);
```
Comment (gustavonihei) [inline arch/risc-v/src/rv64gc/riscv_assert.c:289]:
```suggestion
  nxsched_foreach(riscv_dump_task, NULL);
```

**Comment by anchao** [comment]
@gustavonihei Thank you! please review again

**Review by gustavonihei — APPROVED** [review]

**Pull request merged**